lxd/device/pci/pci.go

```diff
@@ -63,7 +63,7 @@ func ParseUeventFile(ueventFilePath string) (Device, error) {
 func DeviceUnbind(pciDev Device) error {
 	driverUnbindPath := fmt.Sprintf("/sys/bus/pci/devices/%s/driver/unbind", pciDev.SlotName)
 	err := ioutil.WriteFile(driverUnbindPath, []byte(pciDev.SlotName), 0600)
-	if err != nil {
+	if err != nil && !os.IsNotExist(err) {
```

Comment:
@monstermunchkin I think I'd be tempted to handle this in the calling code rather than in the function itself, as this is a generic PCI device helper function, we can't assume that all users of it (current and future) would find it acceptable to silently ignore errors if the device is already unbound. Because the err is returned wrapped, we can use `errors.Cause()` to get the original error and compare it.

Reply:
Hmm, I don't know about that one, do we have any caller of `DeviceUnbind` where we shouldn't have this check in place?

I consider us not doing this check in the NIC part of the code to be a bug as if I use a system where I disable the automatic bind of VFs on the PF, then LXD will similarly fail.

Reply:
Fair enough about the `sriov` NIC type needing to detect and ignore this scenario in its `setupSriovParent` function (as right now it assumes that it is bound to the host).

I don't think there's anywhere we don't need it currently, it was more on the principle of less surprising behaviour. For example I could see any scenario where a user of this function is passing in a missing PCI device accidentally (perhaps an empty string) and it would silently succeed giving the impression we had unbound it when infact we were leaving it bound.

Maybe we should expand this function to check that in the case of getting an os.IsNotExist() error that we also check that `/sys/bus/pci/devices/%s` exists, to ensure that the PCI device itself exists, its just that the driver is unbound (as requested). In that case returning nil makes sense, but if it doesn't exist we should return an error to flag up something unexpected is happening.

What do you think?

Reply:
Yeah, checking that the device itself exists seems reasonable.

Reply:
@tomponline actually, the function takes a `pci.Device` as input which AFAICT we can only get for a valid PCI device, so is this still a concern then?

Reply:
@stgraber it makes it less likely, but its just a struct so could have been created from a DB volatile key or similar that may be out of date or incorrectly stored.